keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,925 @@
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+### Exporting Services
+
+Services will not be visible to other clusters in the supercluster by default.
+They must be explicitly marked for export by the user. This allows users to
+decide exactly which services should be visible outside of the local cluster.
+
+Tooling may (and likely will, in the future) be built on top of this to simplify
+the user experience. Some initial ideas are to allow users to specify that all
+services in a given namespace or in a namespace selector or even a whole cluster
+should be automatically exported by default. In that case, a `ServiceExport`
+could be automatically created for all `Services`. This tooling will be designed
+in a separate doc, and is secondary to the main API proposed here.
+
+To mark a service for export to the supercluster, a user will create a
+ServiceExport CR:
+
+```golang
+// ServiceExport declares that the associated service should be exported to
+// other clusters.
+type ServiceExport struct {
+        metav1.TypeMeta `json:",inline"`
+        // +optional
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        // +optional
+        Status ServiceExportStatus `json:"status,omitempty"`
+}
+
+// ServiceExportStatus contains the current status of an export.
+type ServiceExportStatus struct {
+        // +optional
+        // +patchStrategy=merge
+        // +patchMergeKey=type
+        // +listType=map
+        // +listMapKey=type
+        Conditions []ServiceExportCondition `json:"conditions,omitempty"`
+}
+
+// ServiceExportConditionType identifies a specific condition.
+type ServiceExportConditionType string
+
+const {
+      // ServiceExportInitialized means the service export has been noticed
+      // by the controller, has passed validation, has appropriate finalizers
+      // set, and any required supercluster resources like the IP have been
+      // reserved
+      ServiceExportInitialized ServiceExportConditionType = "Initialized"
+      // ServiceExportExported means that the service referenced by this
+      // service export has been synced to all clusters in the supercluster
+      ServiceExportExported ServiceExportConditionType = "Exported"
+}
+
+// ServiceExportCondition contains details for the current condition of this
+// service export.
+//
+// Once [#1624](https://github.com/kubernetes/enhancements/pull/1624) is
+// merged, this will be replaced by metav1.Condition.
+type ServiceExportCondition struct {
+        Type ServiceExportConditionType `json:"type"`
+        // Status is one of {"True", "False", "Unknown"}
+        Status corev1.ConditionStatus `json:"status"`
+        // +optional
+        LastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty"`
+        // +optional
+        Reason *string `json:"reason,omitempty"`
+        // +optional
+        Message *string `json:"message,omitempty"`
```

Comment:
@JeremyOT Should `Message` contain list of clusters that were successfully synced? Won't be better to have a list of clusters synced and yet to be synced?

Reply:
The not-yet-synced clusters would make sense to list in the message, as a human operator might use them to help track down a failure. For synced clusters, the implementation/registry have the list of available clusters, and a condition representing a successful sync would mean that all clusters are synced so I'm not sure what added benefit would come from listing them. I'd worry that we could end up having to carry a large number of cluster names on every export status in the happy path.